lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -267,162 +269,52 @@ def configure(conf)
-        elsif @detect_subservice
-          # Check for specialized GCE environments.
-          # TODO: Add config options for these to allow for running outside GCE?
-          attributes = fetch_gce_metadata('instance/attributes/').split
-          # Do nothing, just don't populate other service's labels.
-          if attributes.include?('gae_backend_name') &&
-             attributes.include?('gae_backend_version')
-            # Managed VM
-            @running_on_managed_vm = true
-            @gae_backend_name =
-                fetch_gce_metadata('instance/attributes/gae_backend_name')
-            @gae_backend_version =
-                fetch_gce_metadata('instance/attributes/gae_backend_version')
-            @resource.type = APPENGINE_CONSTANTS[:resource_type]
-            @resource.labels['module_id'] = @gae_backend_name
-            @resource.labels['version_id'] = @gae_backend_version
-          elsif attributes.include?('kube-env')
-            # Kubernetes/Container Engine
-            @resource.type = CONTAINER_CONSTANTS[:resource_type]
-            @raw_kube_env = fetch_gce_metadata('instance/attributes/kube-env')
-            @kube_env = YAML.load(@raw_kube_env)
-            @resource.labels['cluster_name'] =
-              cluster_name_from_kube_env(@kube_env)
-            detect_cloudfunctions(attributes)
-          elsif attributes.include?('dataproc-cluster-uuid') &&
-                attributes.include?('dataproc-cluster-name')
-            # Dataproc
-            @resource.type = DATAPROC_CONSTANTS[:resource_type]
-            @resource.labels['cluster_uuid'] =
-              fetch_gce_metadata('instance/attributes/dataproc-cluster-uuid')
-            @resource.labels['cluster_name'] =
-              fetch_gce_metadata('instance/attributes/dataproc-cluster-name')
-            @resource.labels['region'] =
-              fetch_gce_metadata('instance/attributes/dataproc-region')
-          end
-        end
-        # Some services have the GCE instance_id and zone as MonitoredResource
-        # labels; for other services we send them as entry labels.
-        if @resource.type == COMPUTE_CONSTANTS[:resource_type] ||
-           @resource.type == CONTAINER_CONSTANTS[:resource_type]
-          @resource.labels['instance_id'] = @vm_id
-          @resource.labels['zone'] = @zone
-        else
-          common_labels["#{COMPUTE_CONSTANTS[:service]}/resource_id"] = @vm_id
-          common_labels["#{COMPUTE_CONSTANTS[:service]}/zone"] = @zone
+      # Retrieve monitored resource.
+      #
+      # TODO(qingling128): After Metadata Agent support is added, try retrieving
+      # the monitored resource from Metadata Agent first.
+      @resource = determine_agent_level_monitored_resource_via_legacy
+
+      # Set regexp that we should match tags against later on. Using a list
+      # instead of a map to ensure order. For example, tags will be matched
+      # against Cloud Functions first, then GKE.
+      @tag_regexp_list = []
```

Comment:
The original [commit](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/commit/44dccaf724841f90116eb28a1bebf0877a1ed5d7) was from 2 years ago.

```
+        if @running_cloudfunctions
 +          # If the current group of entries is coming from a Cloud Functions
 +          # function, the function name can be extracted from the tag.
 +          match_data = @cloudfunctions_tag_regexp.match(tag)
 +          if match_data
 +            # Service name is set to Cloud Functions only for logs actually
 +            # coming from a function.
 +            @service_name = CLOUDFUNCTIONS_SERVICE
 +            labels = write_log_entries_request['commonLabels']
 +            labels["#{CLOUDFUNCTIONS_SERVICE}/function_name"] =
 +              match_data['function_name']
 +          else
 +            # Other logs are considered as coming from the Container Engine
 +            # service.
 +            @service_name = CONTAINER_SERVICE
 +          end
 +        end
```

`Cloud Functions` is only supported on `GKE`. So we default back to the `GKE` resource type when the tag does not match with `Cloud Function` format. Then we match the tag against `GKE` regex.

I think this makes sense because `@running_cloudfunctions` is true as long as `gcf_region` is available from metadata server. Yet we still want other GKE system logs to be tagged properly.

Also these two tests (`test_multiple_cloudfunctions_logs_tag_not_matched` and `test_multiple_cloudfunctions_logs_tag_not_matched`) were added in this second [commit](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/commit/cb0c64257586a44d50cafc3c1e9ca969c9f1988c) to explicitly ensure the behavior.

WDYT?